tonic/src/transport/endpoint.rs

```diff
@@ -157,7 +157,54 @@ impl Endpoint {
 
     /// Create a channel from this config.
     pub async fn connect(&self) -> Result<Channel, super::Error> {
-        Channel::connect(self.clone()).await
+        // Backwards API compatibility.
+        // Uses TCP if the TLS feature is not enabled, and TLS otherwise.
+
+        #[cfg(feature = "tls")]
+        let connector = super::service::connector(self.tls.clone());
+
+        #[cfg(not(feature = "tls"))]
+        let connector = super::service::connector();
+
+        self.connect_with_connector(connector).await
+    }
+
+    /// Create a channel using a custom connector.
+    ///
+    /// The [`tower_make::MakeConnection`] requirement is an alias for `tower::Service<Uri, Response = AsyncRead +
+    /// Async Write>` - for example, a TCP stream as in [`Endpoint::connect`] above.
+    ///
+    /// # Example
+    /// ```rust
+    /// use hyper::client::connect::HttpConnector;
+    /// use tonic::transport::Endpoint;
+    ///
+    /// // note: This connector is the same as the default provided in `connect()`.
+    /// let mut connector = HttpConnector::new();
+    /// connector.enforce_http(false);
+    /// connector.set_nodelay(true);
+    ///
+    /// let endpoint = Endpoint::from_static("http://example.com");
+    /// endpoint.connect_with_connector(connector); //.await
+    /// ```
+    ///
+    /// # Example with non-default Connector
+    /// ```rust
+    /// // Use for unix-domain sockets
+    /// use hyper_unix_connector::UnixClient;
+    /// use tonic::transport::Endpoint;
+    ///
+    /// let endpoint = Endpoint::from_static("http://example.com");
+    /// endpoint.connect_with_connector(UnixClient); //.await
+    /// ```
+    pub async fn connect_with_connector<C>(&self, connector: C) -> Result<Channel, super::Error>
```

Comment:
so I think it makes more sense for the endpoint to contain a connector option that by default uses the built in tonic one. So instead of providing it at connect time its provided with the endpoint. This means in a load balance list of endpoints they can each have their own type of connector. This should also simplify a lot of the code too.

Reply:
I have poked around at this idea a bit, and it probably involves making several things `pub`, like `transport::service::Connector`, which in turn would require exposing `transport::service::io::BoxedIo`, etc. 

This is because having `Endpoint<C: MakeConnection<...>>` means we will have
```rust
pub fn tls_config(self, tls_config: ClientTlsConfig) -> Endpoint<transport::service::Connector> {
```

and `transport::service::Connector` is currently `pub(crate)`.

What is your plan regarding the `tls` feature? Things might be simpler if it was in its own crate - then usage would just be

```rust
Endpoint::from(...).connector(/* tls connector from the external crate */).connect().await
```

with a plain `HTTPConnector` as the default.

Reply:
You can use a `Box<dyn Connector>` type to avoid the generic for now.